toolchain/lex/tokenized_buffer.cpp

```diff
@@ -352,32 +376,92 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
       emitter_.Emit(source_text.begin() + position + 2,
                     NoWhitespaceAfterCommentIntroducer);
     }
 
-    // Use the current line info to jump to the end of the line.
-    position = current_line_info()->start + current_line_info()->length;
-    // This may be the end of the file in which case we immediately return.
-    if (position == static_cast<ssize_t>(source_text.size())) {
-      // Finished lexing.
-      return;
+    // Skip over this line.
+    ssize_t line_index = line_index_;
+    ++line_index;
+    position = buffer_.line_infos_[line_index].start;
+
+    // A very common pattern is a long block of comment lines all with the same
+    // indent and comment start. We skip these comment blocks in bulk both for
+    // speed and to reduce redundant diagnostics if each line has the same
+    // erroneous comment start like `//!`.
+    //
+    // When we have SIMD support this is even more important for speed, as short
+    // indents can be scanned extremely quickly with SIMD and we expect these to
+    // be the dominant cases.
+    //
+    // TODO: We should extend this to 32-byte SIMD on platforms with support.
+    constexpr int MaxIndent = 13;
+    const int indent = line_info->indent;
+    const ssize_t first_line_start = line_info->start;
+    auto skip_to_next_line = [this, indent, &line_index, &position] {
+      // We're guaranteed to have a line here even on a comment on the last line
+      // as we ensure there is an empty line structure at the end of every file.
+      ++line_index;
+      auto* next_line_info = &buffer_.line_infos_[line_index];
+      next_line_info->indent = indent;
+      position = next_line_info->start;
+    };
+    if (CARBON_USE_SIMD &&
+        position + 16 < static_cast<ssize_t>(source_text.size()) &&
+        indent <= MaxIndent) {
+#if __x86_64__
+      // Load a mask based on the amount of text we want to compare.
+      auto mask = prefix_masks[indent + 3];
+      // And use the current line's prefix as the exemplar to compare against.
+      // We don't mask here as we will mask when doing the comparison.
+      auto prefix = _mm_loadu_si128(reinterpret_cast<const __m128i*>(
+          source_text.data() + first_line_start));
+      do {
+        // Load the next line to consider's prefix.
+        auto next_prefix = _mm_loadu_si128(
+            reinterpret_cast<const __m128i*>(source_text.data() + position));
+        // Compute the difference between the next line and our exemplar. Again,
+        // we don't mask the difference because the comparison below will be
+        // masked.
+        auto prefix_diff = _mm_xor_si128(prefix, next_prefix);
+        // If we have any differences (non-zero bits) within the mask, we can't
+        // skip the next line too.
+        if (!_mm_test_all_zeros(mask, prefix_diff)) {
+          break;
+        }
+
+        skip_to_next_line();
+      } while (position + 16 < static_cast<ssize_t>(source_text.size()));
+#endif
```

Comment:
```suggestion
#elif CARBON_USE_SIMD
#error Unknown target for SIMD comment skipping.
#endif
```
Just in case this `#if` and the one earlier in the file get out of sync.